File: android/app/src/main/java/com/usblanbridge/core/PrintHistory.java
package com.usblanbridge.core;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * A record of every job the phone handled: when, from which device, how it arrived, how big, whether it printed,
 * and a readable extract of the text that reached the printer.
 *
 * Mirrors the Windows PrintHistory so both builds answer the same question the same way. Rows are appended to a
 * daily CSV next to the log file, which can be opened on the phone or pulled over adb.
 */
public final class PrintHistory {

    public static final class Record {
        public Date time;
        public String source;
        public String printer;
        public String path;
        public long bytes;
        public String status;
        /** Readable text, flattened to one line for the CSV. */
        public String preview;
        /** The ticket as it would look on paper, line by line, with markers for images, barcodes, QR codes and the cut. */
        public String ticket;

        public boolean failed() {
            return status == null || !status.equalsIgnoreCase("Printed");
        }

        public String timeText() {
            return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).format(time);
        }
    }

    public interface Listener {
        void onRecord(Record record);
    }

    private static final int MAX_IN_MEMORY = 500;
    private static final int PREVIEW_CHARS = 400;

    private static final List<Record> RECORDS = new ArrayList<>();
    private static final List<Listener> LISTENERS = new ArrayList<>();

    private PrintHistory() {
    }

    public static List<Record> snapshot() {
        synchronized (RECORDS) {
            return new ArrayList<>(RECORDS);
        }
    }

    public static int count() {
        synchronized (RECORDS) {
            return RECORDS.size();
        }
    }

    public static void clear() {
        synchronized (RECORDS) {
            RECORDS.clear();
        }
    }

    public static void addListener(Listener l) {
        synchronized (RECORDS) {
            LISTENERS.add(l);
        }
    }

    public static void removeListener(Listener l) {
        synchronized (RECORDS) {
            LISTENERS.remove(l);
        }
    }

    public static Record add(String source, String printer, String path, byte[] data, int length, String status) {
        Record r = new Record();
        r.time = new Date();
        r.source = source == null ? "?" : source;
        r.printer = printer == null ? "?" : printer;
        r.path = path == null ? "" : path;
        r.bytes = length;
        r.status = status == null ? "Printed" : status;
        r.preview = extractText(data, length, PREVIEW_CHARS);
        try {
            r.ticket = TicketText.render(data, length);
        } catch (Throwable t) {
            r.ticket = "";
        }

        List<Listener> copy;
        synchronized (RECORDS) {
            RECORDS.add(r);
            while (RECORDS.size() > MAX_IN_MEMORY) RECORDS.remove(0);
            copy = new ArrayList<>(LISTENERS);
        }

        appendCsv(r);

        for (Listener l : copy) {
            try {
                l.onRecord(r);
            } catch (Throwable ignored) {
            }
        }
        return r;
    }

    /**
     * Pulls the human-readable part out of an ESC/POS stream. Control bytes and escape sequences are dropped,
     * so what remains is roughly the text printed on the receipt.
     */
    public static String extractText(byte[] data, int length, int maxChars) {
        if (data == null || length <= 0) return "";
        StringBuilder sb = new StringBuilder();
        boolean lastWasSpace = false;

        for (int i = 0; i < length && sb.length() < maxChars; i++) {
            int b = data[i] & 0xFF;

            if (b == 0x1B || b == 0x1D) {
                i += skipCommand(data, length, i);
                continue;
            }

            if (b == 0x0A || b == 0x0D) {
                if (!lastWasSpace) {
                    sb.append(' ');
                    lastWasSpace = true;
                }
                continue;
            }

            if (b >= 0x20 && b <= 0x7E) {
                char c = (char) b;
                if (c == ' ') {
                    if (lastWasSpace) continue;
                    lastWasSpace = true;
                } else {
                    lastWasSpace = false;
                }
                sb.append(c);
            } else if (b >= 0xA0) {
                sb.append((char) b);
                lastWasSpace = false;
            }
        }

        String text = sb.toString().trim();
        if (text.length() >= maxChars) text += "...";
        return text;
    }

    /** Extra bytes to skip for a command starting at i. Conservative, not exhaustive. */
    private static int skipCommand(byte[] data, int length, int i) {
        int lead = data[i] & 0xFF;
        int next = (i + 1 < length) ? (data[i + 1] & 0xFF) : 0;

        if (lead == 0x1B) {
            switch (next) {
                case 0x40:
                    return 1;
                case 0x61: case 0x4D: case 0x2D:
                case 0x45: case 0x47: case 0x7B:
                case 0x33: case 0x4A: case 0x64:
                case 0x74: case 0x56: case 0x21:
                    return 2;
                case 0x70:
                    return 4;
                default:
                    return 1;
            }
        }

        switch (next) {
            case 0x21: case 0x42: case 0x62:
            case 0x48: case 0x66: case 0x77:
            case 0x68: case 0x72: case 0x49:
            case 0x61: case 0x6B:
                return 2;
            case 0x56:
                return 3;
            default:
                return 1;
        }
    }

    private static synchronized void appendCsv(Record r) {
        File dir = Log.getDirectory();
        if (dir == null) return;
        try {
            String day = new SimpleDateFormat("yyyyMMdd", Locale.US).format(r.time);
            File file = new File(dir, "prints-" + day + ".csv");
            boolean isNew = !file.exists();
            Writer writer = new OutputStreamWriter(new FileOutputStream(file, true), "UTF-8");
            try {
                if (isNew) writer.write("Time,Source,Printer,Path,Bytes,Status,Text\n");
                writer.write(csv(r.timeText()) + "," + csv(r.source) + "," + csv(r.printer) + ","
                        + csv(r.path) + "," + r.bytes + "," + csv(r.status) + "," + csv(r.preview) + "\n");
            } finally {
                writer.close();
            }
        } catch (Throwable ignored) {
        }
    }

    private static String csv(String value) {
        if (value == null) return "\"\"";
        return "\"" + value.replace("\"", "\"\"") + "\"";
    }
}

File: src/UsbLanPrinterBridge/Core/PrintHistory.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;

namespace UsbLanPrinterBridge.Core
{
    /// <summary>One completed (or failed) print job.</summary>
    public sealed class PrintRecord
    {
        public DateTime Time { get; set; }
        /// <summary>Address of the device that sent the job.</summary>
        public string Source { get; set; }
        public string Printer { get; set; }
        /// <summary>How it arrived: "raw 9100", "ePOS http", "ePOS https", "test".</summary>
        public string Path { get; set; }
        public long Bytes { get; set; }
        /// <summary>"Printed", or the reason it did not.</summary>
        public string Status { get; set; }
        /// <summary>Readable text pulled out of the ESC/POS stream, flattened to one line for the grid and the CSV.</summary>
        public string Preview { get; set; }
        /// <summary>The ticket as it would look on paper, line by line, with markers for images, barcodes, QR codes and the cut.</summary>
        public string Ticket { get; set; }

        public bool Failed { get { return !string.Equals(Status, "Printed", StringComparison.OrdinalIgnoreCase); } }

        public string TimeText { get { return Time.ToString("yyyy-MM-dd HH:mm:ss"); } }
    }

    /// <summary>
    /// Keeps a record of every job the bridge handled: when, from whom, how big, and a readable extract of the
    /// text that went to the printer. Held in memory for the window and appended to a daily CSV so it survives
    /// restarts.
    ///
    /// Raw dumps can be enabled to write the exact bytes of each job to disk, which is what you want when a
    /// client is rejected and the question is what it actually sent.
    /// </summary>
    public static class PrintHistory
    {
        private const int MaxInMemory = 1000;
        private const int PreviewChars = 400;

        private static readonly object Gate = new object();
        private static readonly List<PrintRecord> Records = new List<PrintRecord>();
        private static string _csvDay;

        /// <summary>When true, the raw bytes of every job are written to logs\jobs for inspection.</summary>
        public static bool SaveRawJobs { get; set; }

        public static event Action<PrintRecord> RecordAdded;

        public static string JobDumpDirectory { get { return Path.Combine(ConfigStore.LogDirectory, "jobs"); } }

        public static List<PrintRecord> Snapshot()
        {
            lock (Gate) return new List<PrintRecord>(Records);
        }

        public static int Count
        {
            get { lock (Gate) return Records.Count; }
        }

        public static void Clear()
        {
            lock (Gate) Records.Clear();
        }

        public static PrintRecord Add(string source, string printer, string path, byte[] data, int length, string status)
        {
            var record = new PrintRecord
            {
                Time = DateTime.Now,
                Source = source ?? "?",
                Printer = printer ?? "?",
                Path = path ?? "",
                Bytes = length,
                Status = status ?? "Printed",
                Preview = ExtractText(data, length, PreviewChars),
                Ticket = SafeRender(data, length)
            };

            lock (Gate)
            {
                Records.Add(record);
                if (Records.Count > MaxInMemory) Records.RemoveRange(0, Records.Count - MaxInMemory);
            }

            AppendCsv(record);
            if (SaveRawJobs && data != null && length > 0) DumpRaw(record, data, length);

            Action<PrintRecord> handler = RecordAdded;
            if (handler != null) { try { handler(record); } catch { } }
            return record;
        }

        private static string SafeRender(byte[] data, int length)
        {
            try { return EscPosTicketText.Render(data, length); }
            catch { return ""; }
        }

        /// <summary>
        /// Pulls the human-readable part out of an ESC/POS stream. Control bytes and escape sequences are
        /// dropped, so what is left is roughly the text the customer sees on the receipt.
        /// </summary>
        public static string ExtractText(byte[] data, int length, int maxChars)
        {
            if (data == null || length <= 0) return "";
            var sb = new StringBuilder();
            bool lastWasSpace = false;

            for (int i = 0; i < length && sb.Length < maxChars; i++)
            {
                byte b = data[i];

                if (b == 0x1B || b == 0x1D) // ESC or GS: skip the command and its most common operands
                {
                    i += SkipCommand(data, length, i);
                    continue;
                }

                if (b == 0x0A || b == 0x0D)
                {
                    if (!lastWasSpace) { sb.Append(' '); lastWasSpace = true; }
                    continue;
                }

                if (b >= 0x20 && b <= 0x7E)
                {
                    char c = (char)b;
                    if (c == ' ')
                    {
                        if (lastWasSpace) continue;
                        lastWasSpace = true;
                    }
                    else lastWasSpace = false;
                    sb.Append(c);
                }
                else if (b >= 0xA0)
                {
                    // Accented characters from a code page; show them rather than dropping the word.
                    sb.Append((char)b);
                    lastWasSpace = false;
                }
            }

            string text = sb.ToString().Trim();
            if (text.Length >= maxChars) text += "...";
            return text;
        }

        /// <summary>Number of extra bytes to skip for a command starting at index i. Conservative, not exhaustive.</summary>
        private static int SkipCommand(byte[] data, int length, int i)
        {
            byte lead = data[i];
            byte next = (i + 1 < length) ? data[i + 1] : (byte)0;

            if (lead == 0x1B)
            {
                switch (next)
                {
                    case 0x40: return 1;                 // ESC @
                    case 0x61: case 0x4D: case 0x2D:     // ESC a / M / -
                    case 0x45: case 0x47: case 0x7B:     // ESC E / G / {
                    case 0x33: case 0x4A: case 0x64:     // ESC 3 / J / d
                    case 0x74: case 0x56: case 0x21:     // ESC t / V / !
                        return 2;
                    case 0x70: return 4;                 // ESC p m t1 t2
                    default: return 1;
                }
            }

            // GS
            switch (next)
            {
                case 0x21: case 0x42: case 0x62:         // GS ! / B / b
                case 0x48: case 0x66: case 0x77:         // GS H / f / w
                case 0x68: case 0x72: case 0x49:         // GS h / r / I
                case 0x61:                                // GS a
                    return 2;
                case 0x56: return 3;                     // GS V m (n)
                case 0x6B: return 2;                     // GS k: data follows, handled as text below
                default: return 1;
            }
        }

        private static void AppendCsv(PrintRecord r)
        {
            try
            {
                string dir = ConfigStore.LogDirectory;
                Directory.CreateDirectory(dir);
                string day = r.Time.ToString("yyyyMMdd");
                string file = Path.Combine(dir, "prints-" + day + ".csv");

                bool isNew = !File.Exists(file);
                using (var writer = new StreamWriter(new FileStream(file, FileMode.Append, FileAccess.Write, FileShare.ReadWrite), new UTF8Encoding(true)))
                {
                    if (isNew || _csvDay != day)
                    {
                        if (isNew) writer.WriteLine("Time,Source,Printer,Path,Bytes,Status,Text");
                        _csvDay = day;
                    }
                    writer.WriteLine(string.Join(",",
                        Csv(r.TimeText), Csv(r.Source), Csv(r.Printer), Csv(r.Path),
                        r.Bytes.ToString(CultureInfo.InvariantCulture), Csv(r.Status), Csv(r.Preview)));
                }
            }
            catch
            {
                // logging must never break printing
            }
        }

        private static string Csv(string value)
        {
            if (value == null) return "\"\"";
            return "\"" + value.Replace("\"", "\"\"") + "\"";
        }

        private static void DumpRaw(PrintRecord r, byte[] data, int length)
        {
            try
            {
                Directory.CreateDirectory(JobDumpDirectory);
                string name = r.Time.ToString("yyyyMMdd-HHmmss-fff") + "_" + Sanitize(r.Source) + ".bin";
                using (var fs = new FileStream(Path.Combine(JobDumpDirectory, name), FileMode.Create, FileAccess.Write))
                {
                    fs.Write(data, 0, length);
                }
            }
            catch
            {
            }
        }

        private static string Sanitize(string s)
        {
            if (string.IsNullOrEmpty(s)) return "unknown";
            var sb = new StringBuilder();
            foreach (char c in s) sb.Append(char.IsLetterOrDigit(c) ? c : '-');
            return sb.ToString();
        }

        /// <summary>Formats bytes as a readable hex dump, used when tracing what a client sent.</summary>
        public static string HexDump(byte[] data, int length, int maxBytes)
        {
            if (data == null || length <= 0) return "(no data)";
            int count = Math.Min(length, maxBytes);
            var sb = new StringBuilder();
            for (int i = 0; i < count; i += 16)
            {
                var hex = new StringBuilder();
                var text = new StringBuilder();
                for (int j = i; j < i + 16 && j < count; j++)
                {
                    hex.Append(data[j].ToString("X2")).Append(' ');
                    text.Append(data[j] >= 0x20 && data[j] <= 0x7E ? (char)data[j] : '.');
                }
                sb.Append(hex.ToString().PadRight(48)).Append(' ').Append(text).AppendLine();
            }
            if (length > count) sb.Append("... ").Append(length - count).Append(" more byte(s)");
            return sb.ToString().TrimEnd();
        }
    }
}
